consensus/core/src/synchronizer.rs

```diff
@@ -0,0 +1,480 @@
+}
+
+impl Synchronizer {
+    pub fn start<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher>(
+        network_client: Arc<C>,
+        context: Arc<Context>,
+        core_dispatcher: Arc<D>,
+        block_verifier: Arc<V>,
+    ) -> Arc<SynchronizerHandle> {
+        let (commands_sender, commands_receiver) = channel(1_000);
+
+        // Spawn the tasks to fetch the blocks from the others
+        let mut fetch_block_senders = BTreeMap::new();
+        let mut tasks = JoinSet::new();
+        for (index, _) in context.committee.authorities() {
+            if index == context.own_index {
+                continue;
+            }
+            let (sender, receiver) = channel(FETCH_BLOCKS_CONCURRENCY);
+            tasks.spawn(Self::fetch_blocks_from_authority(
+                index,
+                network_client.clone(),
+                block_verifier.clone(),
+                context.clone(),
+                core_dispatcher.clone(),
+                receiver,
+            ));
+            fetch_block_senders.insert(index, sender);
+        }
+
+        // Spawn the task to listen to the requests & periodic runs
+        tasks.spawn(async {
+            let mut s = Self {
+                context,
+                commands_receiver,
+                fetch_block_senders,
+            };
+            s.run().await;
+        });
+
+        Arc::new(SynchronizerHandle {
+            commands_sender,
+            tasks: Mutex::new(tasks),
+        })
+    }
+
+    async fn fetch_blocks_from_authority<
+        C: NetworkClient,
+        V: BlockVerifier,
+        D: CoreThreadDispatcher,
+    >(
+        peer_index: AuthorityIndex,
+        network_client: Arc<C>,
+        block_verifier: Arc<V>,
+        context: Arc<Context>,
+        core_dispatcher: Arc<D>,
+        mut receiver: Receiver<BTreeSet<BlockRef>>,
+    ) {
+        const REQUEST_TIMEOUT: Duration = Duration::from_millis(500);
+        const MAX_RETRIES: u32 = 5;
+
+        let mut requests = FuturesUnordered::new();
+
+        loop {
+            tokio::select! {
+                Some(block_refs) = receiver.recv(), if requests.len() < FETCH_BLOCKS_CONCURRENCY => {
+                    requests.push(Self::fetch_blocks_request(network_client.clone(), peer_index, block_refs, REQUEST_TIMEOUT, 1))
+                },
+                Some((response, block_refs, retries)) = requests.next() => {
+                    match response {
+                        Ok(Ok(blocks)) => {
+                            if let Err(err) = Self::process_fetched_blocks(blocks,
+                                peer_index,
+                                block_refs,
+                                core_dispatcher.clone(),
+                                block_verifier.clone(),
+                                context.clone()).await {
+                                error!("Error while processing fetched blocks from peer {peer_index}: {err}");
```

Comment:
`error` should be reserved for logical bugs.

```suggestion
                                warn!("Error while processing fetched blocks from peer {peer_index}: {err}");
```

Reply:
agreed